src/utility/Auth.js

```diff
@@ -0,0 +1,57 @@
+import React, { createContext, useContext, useState, useEffect } from 'react'
+import firebase from 'firebase/app'
+import 'firebase/auth'
+
+const AuthContext = createContext()
+
+export function useAuth() {
+  return useContext(AuthContext)
+}
+
+export function AuthProvider({ children }) {
+  const [user, setUser] = useState(null)
+
+  useEffect(() => {
+    return firebase.auth().onAuthStateChanged(async currUser => {
+      setUser(currUser)
+    })
+  })
+
+  return (
+    <AuthContext.Provider value={{ isAuthed: !!user, user, setUser }}>
+      {children}
+    </AuthContext.Provider>
+  )
+}
+
+export const googleSignIn = async (setUser, setLocation) => {
+  await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
+  const provider = new firebase.auth.GoogleAuthProvider()
+  try {
+    await firebase.auth().signInWithPopup(provider)
+    const user = firebase.auth().currentUser
+    setUser(user)
+    setLocation('/application')
+    return null
+  } catch (e) {
+    return e
+  }
+}
+
+export const githubSignIn = async (setUser, setLocation) => {
+  await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
+  const provider = new firebase.auth.GithubAuthProvider()
+  try {
+    await firebase.auth().signInWithPopup(provider)
+    const user = firebase.auth().currentUser
+    setUser(user)
+    setLocation('/application')
```

Comment:
According to Tiffany, we would take the user to the app form if they haven't started or submitted their application yet. So I think we will need to get the user's application status first?

Reply:
not a bad idea, we could route them based on application status (if accepted, they'll be routed to judging instead of the application)

Reply:
would it be possible to pass the user app status in the AuthContext too? the sidebar display will depend on it

Reply:
Yeah its totally possible, I'll add it

Reply:
Actually this seems like out of scope for this PR because we also need to create the hacker app document on first sign in. For now, I think this PR has a big enough scope.